File: 2024/day23/node/solution.js
import { readFileSync } from 'fs';

function parseInput(filename) {
    const graph = new Map();
    const lines = readFileSync(filename, 'utf8').trim().split('\n');

    for (const line of lines) {
        const [a, b] = line.split('-');
        if (!graph.has(a)) graph.set(a, new Set());
        if (!graph.has(b)) graph.set(b, new Set());
        graph.get(a).add(b);
        graph.get(b).add(a);
    }
    return graph;
}

function findTriangles(graph) {
    const triangles = new Set();

    for (const [a, neighborsA] of graph) {
        for (const b of neighborsA) {
            if (a < b) {
                const neighborsB = graph.get(b);
                // Find common neighbors
                for (const c of neighborsA) {
                    if (neighborsB.has(c)) {
                        // Create sorted tuple key
                        const tri = [a, b, c].sort().join(',');
                        triangles.add(tri);
                    }
                }
            }
        }
    }
    return triangles;
}

function part1(graph) {
    const triangles = findTriangles(graph);
    let count = 0;
    for (const tri of triangles) {
        const nodes = tri.split(',');
        if (nodes.some(n => n.startsWith('t'))) {
            count++;
        }
    }
    return count;
}

function bronKerbosch(graph, r, p, x, cliques) {
    if (p.size === 0 && x.size === 0) {
        cliques.push(new Set(r));
        return;
    }

    // Use pivot to reduce branching
    const px = new Set([...p, ...x]);
    let pivot = null;
    let maxNeighbors = -1;
    for (const v of px) {
        const neighbors = graph.get(v);
        const count = [...p].filter(n => neighbors.has(n)).length;
        if (count > maxNeighbors) {
            maxNeighbors = count;
            pivot = v;
        }
    }

    const pivotNeighbors = graph.get(pivot);
    const candidates = [...p].filter(v => !pivotNeighbors.has(v));

    for (const v of candidates) {
        const neighbors = graph.get(v);
        const newR = new Set(r);
        newR.add(v);
        const newP = new Set([...p].filter(n => neighbors.has(n)));
        const newX = new Set([...x].filter(n => neighbors.has(n)));

        bronKerbosch(graph, newR, newP, newX, cliques);

        p.delete(v);
        x.add(v);
    }
}

function part2(graph) {
    const cliques = [];
    const allNodes = new Set(graph.keys());
    bronKerbosch(graph, new Set(), allNodes, new Set(), cliques);

    // Find largest clique
    let largest = cliques[0];
    for (const clique of cliques) {
        if (clique.size > largest.size) {
            largest = clique;
        }
    }

    return [...largest].sort().join(',');
}

const graph = parseInput('../input.txt');
console.log('Part 1:', part1(graph));
console.log('Part 2:', part2(graph));

File: 2024/day23/rust/solution.rs
use std::collections::{HashMap, HashSet};
use std::fs;

fn parse_input(filename: &str) -> HashMap<String, HashSet<String>> {
    let content = fs::read_to_string(filename).expect("Failed to read input file");
    let mut graph: HashMap<String, HashSet<String>> = HashMap::new();

    for line in content.lines() {
        let parts: Vec<&str> = line.trim().split('-').collect();
        if parts.len() == 2 {
            let a = parts[0].to_string();
            let b = parts[1].to_string();

            graph.entry(a.clone()).or_insert_with(HashSet::new).insert(b.clone());
            graph.entry(b).or_insert_with(HashSet::new).insert(a);
        }
    }

    graph
}

fn find_triangles(graph: &HashMap<String, HashSet<String>>) -> HashSet<Vec<String>> {
    let mut triangles = HashSet::new();

    for a in graph.keys() {
        if let Some(neighbors_a) = graph.get(a) {
            for b in neighbors_a {
                if a < b {  // Only process each edge once
                    // Find common neighbors
                    if let Some(neighbors_b) = graph.get(b) {
                        let common: HashSet<_> = neighbors_a.intersection(neighbors_b).collect();
                        for c in common {
                            // Create sorted triangle to avoid duplicates
                            let mut tri = vec![a.clone(), b.clone(), (*c).clone()];
                            tri.sort();
                            triangles.insert(tri);
                        }
                    }
                }
            }
        }
    }

    triangles
}

fn part1(graph: &HashMap<String, HashSet<String>>) -> usize {
    let triangles = find_triangles(graph);

    triangles.iter()
        .filter(|tri| tri.iter().any(|node| node.starts_with('t')))
        .count()
}

fn bron_kerbosch(
    graph: &HashMap<String, HashSet<String>>,
    r: HashSet<String>,
    mut p: HashSet<String>,
    mut x: HashSet<String>,
    cliques: &mut Vec<HashSet<String>>,
) {
    if p.is_empty() && x.is_empty() {
        cliques.push(r);
        return;
    }

    // Use pivot to reduce branching
    let union: HashSet<_> = p.union(&x).cloned().collect();
    let pivot = union.iter()
        .max_by_key(|v| {
            graph.get(*v)
                .map(|neighbors| neighbors.intersection(&p).count())
                .unwrap_or(0)
        })
        .cloned();

    // Get pivot's neighbors
    let pivot_neighbors = if let Some(piv) = pivot {
        graph.get(&piv).cloned().unwrap_or_default()
    } else {
        HashSet::new()
    };

    // Process vertices not in pivot's neighborhood
    let candidates: Vec<_> = p.difference(&pivot_neighbors).cloned().collect();

    for v in candidates {
        let mut new_r = r.clone();
        new_r.insert(v.clone());

        let neighbors = graph.get(&v).cloned().unwrap_or_default();
        let new_p: HashSet<_> = p.intersection(&neighbors).cloned().collect();
        let new_x: HashSet<_> = x.intersection(&neighbors).cloned().collect();

        bron_kerbosch(graph, new_r, new_p, new_x, cliques);

        p.remove(&v);
        x.insert(v);
    }
}

fn part2(graph: &HashMap<String, HashSet<String>>) -> String {
    let mut cliques = Vec::new();
    let all_nodes: HashSet<String> = graph.keys().cloned().collect();

    bron_kerbosch(
        graph,
        HashSet::new(),
        all_nodes,
        HashSet::new(),
        &mut cliques
    );

    // Find the largest clique
    let largest = cliques.iter()
        .max_by_key(|c| c.len())
        .expect("No cliques found");

    // Return sorted, comma-joined password
    let mut nodes: Vec<_> = largest.iter().cloned().collect();
    nodes.sort();
    nodes.join(",")
}

fn main() {
    let graph = parse_input("../input.txt");

    println!("Part 1: {}", part1(&graph));
    println!("Part 2: {}", part2(&graph));
}
